Crash with imageConfig

I run hugo 0.18 on 64bit Linux and use the precompiled executable from the releases page on Github. When I try the new imageConfig function, hugo crashes directly.

My shortcode:
```
{{ with (imageConfig (printf "static/%s" (.Get 0))) }}
	<div itemprop="width" itemscope itemtype="https://schema.org/QuantitativeValue">
            <meta itemprop="unitText" content="px">
            <meta itemprop="value" content="{{ .Width }}">
	</div>
{{ end }}
```


Hugo output:
```
Started building sites ...
panic: assignment to entry in nil map [recovered]
 panic: assignment to entry in nil map

goroutine 129 [running]:
panic(0xa49aa0, 0xc42038c9c0)
 /usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:500 +0x1a1
text/template.errRecover(0xc420581810)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:140 +0x2ad
panic(0xa49aa0, 0xc42038c9c0)
 /usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:458 +0x243
github.com/spf13/hugo/tpl.imageConfig(0xa195e0, 0xc4202faa68, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
 /Users/bep/go/src/github.com/spf13/hugo/tpl/template_funcs.go:425 +0x321
reflect.Value.call(0xa350a0, 0xb5e328, 0x13, 0xae26cd, 0x4, 0xc42041a1c0, 0x1, 0x1, 0xdff4c0, 0xa4eb80, ...)
 /usr/local/Cellar/go/1.7.4/libexec/src/reflect/value.go:434 +0x5c8
reflect.Value.Call(0xa350a0, 0xb5e328, 0x13, 0xc42041a1c0, 0x1, 0x1, 0xdfa940, 0xc4202158b0, 0xa195e0)
 /usr/local/Cellar/go/1.7.4/libexec/src/reflect/value.go:302 +0xa4
text/template.(*state).evalCall(0xc420581790, 0xa72c40, 0xc4202c6180, 0x16, 0xa350a0, 0xb5e328, 0x13, 0xdfa640, 0xc420422570, 0xc4201f4e04, ...)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:658 +0x530
text/template.(*state).evalFunction(0xc420581790, 0xa72c40, 0xc4202c6180, 0x16, 0xc4204225a0, 0xdfa640, 0xc420422570, 0xc4202b5680, 0x2, 0x2, ...)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:530 +0x195
text/template.(*state).evalCommand(0xc420581790, 0xa72c40, 0xc4202c6180, 0x16, 0xc420422570, 0x0, 0x0, 0x0, 0xc4200fa040, 0xc42047ab7f, ...)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:427 +0x6f7
text/template.(*state).evalPipeline(0xc420581790, 0xa72c40, 0xc4202c6180, 0x16, 0xc420215860, 0x0, 0x410126, 0xc4205813e8)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:400 +0xee
text/template.(*state).evalCommand(0xc420581790, 0xa72c40, 0xc4202c6180, 0x16, 0xc420422540, 0x0, 0x0, 0x0, 0xc4205814d8, 0xc4205814d8, ...)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:430 +0x9b8
text/template.(*state).evalPipeline(0xc420581790, 0xa72c40, 0xc4202c6180, 0x16, 0xc420215810, 0xc420032400, 0x44fe20, 0xc420032400)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:400 +0xee
text/template.(*state).walkIfOrWith(0xc420581790, 0x13, 0xa72c40, 0xc4202c6180, 0x16, 0xc420215810, 0xc42036cea0, 0x0)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:255 +0xc4
text/template.(*state).walk(0xc420581790, 0xa72c40, 0xc4202c6180, 0x16, 0xdfab80, 0xc4202cd8c0)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:245 +0x37d
text/template.(*state).walk(0xc420581790, 0xa72c40, 0xc4202c6180, 0x16, 0xdfa820, 0xc42036ccc0)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:234 +0x13d
text/template.(*Template).execute(0xc4202cd5c0, 0xdf0be0, 0xc420328000, 0xa72c40, 0xc4202c6180, 0x0, 0x0)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:189 +0x204
text/template.(*Template).Execute(0xc4202cd5c0, 0xdf0be0, 0xc420328000, 0xa72c40, 0xc4202c6180, 0xc400000001, 0xc400000001)
 /usr/local/Cellar/go/1.7.4/libexec/src/text/template/exec.go:175 +0x53
html/template.(*Template).Execute(0xc42036cc60, 0xdf0be0, 0xc420328000, 0xa72c40, 0xc4202c6180, 0xaa2c80, 0xc4205f0908)
 /usr/local/Cellar/go/1.7.4/libexec/src/html/template/template.go:104 +0x78
github.com/spf13/hugo/hugolib.renderShortcodeWithPage(0xc42036cc60, 0xc4202c6180, 0x0, 0x0)
 /Users/bep/go/src/github.com/spf13/hugo/hugolib/shortcode.go:585 +0xf2
github.com/spf13/hugo/hugolib.renderShortcode(0xc42047ab1e, 0x8, 0x0, 0x0, 0x0, 0xa110e0, 0xc4201ccd20, 0x0, 0x0, 0x0, ...)
 /Users/bep/go/src/github.com/spf13/hugo/hugolib/shortcode.go:281 +0x5d2
github.com/spf13/hugo/hugolib.renderShortcodes.func1(0xa45520, 0xc42039a630, 0xc4205f0d10, 0x0)
 /Users/bep/go/src/github.com/spf13/hugo/hugolib/shortcode.go:328 +0x74
github.com/spf13/hugo/hugolib.executeShortcodeFuncMap(0xc42039a630, 0xaf83b7, 0x1b, 0xc4205f0e30)
 /Users/bep/go/src/github.com/spf13/hugo/hugolib/shortcode.go:309 +0xe5
github.com/spf13/hugo/hugolib.handleShortcodes(0xc420464000, 0xdfd000, 0xc4201a74a0, 0xc420395600, 0xa5e, 0xb00, 0xc420230701, 0x517a19, 0x1, 0xc420230798, ...)
 /Users/bep/go/src/github.com/spf13/hugo/hugolib/hugo_sites.go:437 +0x194
github.com/spf13/hugo/hugolib.(*Site).preparePagesForRender.func1(0xc420357140, 0xc4201dca00, 0xc4200ca3c0, 0xc420519ba0)
 /Users/bep/go/src/github.com/spf13/hugo/hugolib/hugo_sites.go:387 +0x185
created by github.com/spf13/hugo/hugolib.(*Site).preparePagesForRender
 /Users/bep/go/src/github.com/spf13/hugo/hugolib/hugo_sites.go:416 +0xe5
```

--- a/tpl/template_funcs.go
+++ b/tpl/template_funcs.go
@@ -378,19 +378,23 @@ func ResetCaches() {
 
 // imageConfigCache is a lockable cache for image.Config objects. It must be
 // locked before reading or writing to config.
-var imageConfigCache struct {
-	sync.RWMutex
+type imageConfigCache struct {
 	config map[string]image.Config
+	sync.RWMutex
+}
+
+var defaultImageConfigCache = imageConfigCache{
+	config: map[string]image.Config{},
 }
 
 // resetImageConfigCache initializes and resets the imageConfig cache for the
 // imageConfig template function. This should be run once before every batch of
 // template renderers so the cache is cleared for new data.
 func resetImageConfigCache() {
-	imageConfigCache.Lock()
-	defer imageConfigCache.Unlock()
+	defaultImageConfigCache.Lock()
+	defer defaultImageConfigCache.Unlock()
 
-	imageConfigCache.config = map[string]image.Config{}
+	defaultImageConfigCache.config = map[string]image.Config{}
 }
 
 // imageConfig returns the image.Config for the specified path relative to the
@@ -406,9 +410,9 @@ func imageConfig(path interface{}) (image.Config, error) {
 	}
 
 	// Check cache for image config.
-	imageConfigCache.RLock()
-	config, ok := imageConfigCache.config[filename]
-	imageConfigCache.RUnlock()
+	defaultImageConfigCache.RLock()
+	config, ok := defaultImageConfigCache.config[filename]
+	defaultImageConfigCache.RUnlock()
 
 	if ok {
 		return config, nil
@@ -421,9 +425,9 @@ func imageConfig(path interface{}) (image.Config, error) {
 
 	config, _, err = image.DecodeConfig(f)
 
-	imageConfigCache.Lock()
-	imageConfigCache.config[filename] = config
-	imageConfigCache.Unlock()
+	defaultImageConfigCache.Lock()
+	defaultImageConfigCache.config[filename] = config
+	defaultImageConfigCache.Unlock()
 
 	return config, err
 }

--- a/tpl/template_funcs_test.go
+++ b/tpl/template_funcs_test.go
@@ -629,6 +629,17 @@ func TestImageConfig(t *testing.T) {
 		input      []byte
 		expected   image.Config
 	}{
+		// Make sure that the cache is initialized by default.
+		{
+			resetCache: false,
+			path:       "a.png",
+			input:      blankImage(10, 10),
+			expected: image.Config{
+				Width:      10,
+				Height:     10,
+				ColorModel: color.NRGBAModel,
+			},
+		},
 		{
 			resetCache: true,
 			path:       "a.png",
@@ -685,8 +696,8 @@ func TestImageConfig(t *testing.T) {
 			t.Errorf("[%d] imageConfig: expected '%v', got '%v'", i, this.expected, result)
 		}
 
-		if len(imageConfigCache.config) == 0 {
-			t.Error("imageConfigCache should have at least 1 item")
+		if len(defaultImageConfigCache.config) == 0 {
+			t.Error("defaultImageConfigCache should have at least 1 item")
 		}
 	}
 
@@ -705,8 +716,8 @@ func TestImageConfig(t *testing.T) {
 	// test cache clearing
 	ResetCaches()
 
-	if len(imageConfigCache.config) != 0 {
-		t.Error("ResetCaches should have cleared imageConfigCache")
+	if len(defaultImageConfigCache.config) != 0 {
+		t.Error("ResetCaches should have cleared defaultImageConfigCache")
 	}
 }
 

